language/move-compiler/src/typing/translate.rs

```diff
@@ -2117,17 +2170,83 @@ impl crate::shared::ast_debug::AstDebug for ExpDotted_ {
 // Calls
 //**************************************************************************************************
 
+fn local_call(
+    context: &mut Context,
+    loc: Loc,
+    var: Var,
+    argloc: Loc,
+    args: Vec<T::Exp>,
+) -> (Type, T::UnannotatedExp_) {
+    let ty = context.get_local(var.0.loc, "function usage", &var);
+    match ty.value {
+        Type_::UnresolvedError => {
+            assert!(context.env.has_diags());
+            (ty, T::UnannotatedExp_::UnresolvedError)
+        }
+        Type_::Apply(_, sp!(_, TypeName_::Builtin(sp!(_, BuiltinTypeName_::Fun))), targs) => {
+            let (arguments, arg_tys) = call_args(
+                context,
+                loc,
+                || format!("Invalid call of '{}'", var),
+                targs.len() - 1,
+                argloc,
+                args,
+            );
+            assert!(arg_tys.len() == targs.len() - 1);
+            for (arg_ty, param_ty) in arg_tys
+                .into_iter()
+                .zip(targs[0..targs.len() - 1].iter().cloned())
+            {
+                let msg = || format!("Invalid call of '{}'. Invalid argument type", var);
+                subtype(context, loc, msg, arg_ty, param_ty);
+            }
+            (
+                targs[targs.len() - 1].clone(),
+                T::UnannotatedExp_::LocalCall(var, arguments),
+            )
+        }
+        ty_ => {
```

Comment:
This might workout... given how restrictive lambdas are but...

This should really be likely something like 
```
let args = make_tvars(...);
let ret_ty = make_tvar(...);
subtype(ty, sp(loc, Type::Lambda(args, ret_ty)), ...);
```
Just a bit more principled and will deal with all sorts of other edge cases for you

Reply:
This will also make the whole `targs[targs.len() - 1].clone()` thing cleaner too

Reply:
I do it like this because in my experience, error messages tend to be better when you break types down in calls, instead of resorting to function type unification. I don't know though whether your type unification error reporting takes alreadycare of this anyway. Another aspect is that I wanted to reuse the `call_args` function also used by `module_call`, so getting all treatment it does.